src/metaopt/worker/experiment.py

```diff
@@ -0,0 +1,231 @@
+           is called.
+
+        :param name: Describe a configuration with a unique identifier per :attr:`user`.
+        :type name: str
+        """
+        self._init_done = False
+        self._db = Database()  # fetch database instance
+
+        self._id = None
+        self.name = name
+        self.refers = None
+        user = getpass.getuser()
+        stamp = datetime.datetime.utcnow()
+        self.metadata = {'user': user, 'datetime': stamp}
+        self.pool_size = None
+        self.max_trials = None
+        self.status = None
+        self.algorithms = None
+
+        config = self._db.read('experiments',
+                               {'name': name, 'metadata.user': user})
+        if config:
+            config = sorted(config, key=lambda x: x['metadata']['datetime'],
+                            reverse=True)[0]
+            #  assert (len(config) == 1) is not True  # currently
+            for varname in self.__slots__:
+                if not varname.startswith('_'):
+                    setattr(self, varname, config[varname])
+            self._id = config['_id']
+
+    def reserve_trial(self, score_handle=None):
+        """Find *new* trials that exist currently in database and select one of
+        them based on the highest score return from `score_handle` callable.
+
+        :param score_handle: A way to decide which trial out of the *new* ones to
+           to pick as *pending*, defaults to a random choice.
+        :type score_handle: callable
+
+        :return: selected `Trial` object
+        """
+        raise NotImplementedError()
+
+    def push_completed_trial(self, trial):
+        """Inform database about an evaluated `trial` with results.
+
+        :param trial: Corresponds to a successful evaluation of a particular run.
+        :type trial: `Trial`
+
+        .. note:: Change status from *pending* to *completed*.
+        """
+        raise NotImplementedError()
+
+    def register_trials(self, trials):
+        """Inform database about *new* suggested trial with specific parameter
+        values. Each of them correspond to a different possible run.
+
+        :type trials: list of `Trial`
+        """
+        raise NotImplementedError()
+
+    @property
+    def is_done(self):
+        """Count how many trials have been completed and compare with `max_trials`.
+
+        .. note:: To be used as a terminating condition in a ``Worker``.
+        """
+        raise NotImplementedError()
+
+    @property
+    def config(self):
+        """Return a copy of an `Experiment` configuration as a dictionary."""
+        if self._init_done:
+            return self._db.read('experiments', {'_id': self._id})[0]
+
+        config = dict()
+        for varname in self.__slots__:
+            if not varname.startswith('_'):
+                config[varname] = getattr(self, varname)
+        return copy.deepcopy(config)
+
+    @config.setter
+    def config(self, config):
```

Comment:
It seems confusing to me that we need to do ```experiment.config = some_config``` in order to consider it _initialized_. 

```python
>>> experiment = Experiment("experiment_name")
>>> print experiment.stats
[...]
RuntimeError: Experiment was not configured.
```
:thinking: 

It's more like a side-effect when we use a setter rather than an explicit method.

Currently it would be this
```python
>>> experiment = Experiment("experiment_name")
>>> experiment.config = some_config
>>> print experiment.stats
{some stats}
```

What would you think about this?
```python
>>> experiment = Experiment("experiment_name")
>>> experiment.configure(some_config)
>>> print experiment.stats
{some stats}
```
Or this?

```python
>>> experiment = Experiment("experiment_name", some_config)
>>> print experiment.stats
{some stats}
```

Reply:
Yes probably creating a function, instead of a property, would be better. I would like to have a getter though. Actually It doesn't seem as bad, to have a getter only for `experiment.configuration` property and a method `experiment.configure(config_from_user)`. I will do that.

Reply:
I agree for the getter, I didn't mean otherwise. ```experiment.config``` and ```experiment.configure(config_from_user)``` seems good to me.